File: python/1128.py
# Link: https://leetcode.com/problems/number-of-equivalent-domino-pairs/description/
# Runtime: 11 ms
# Memory: 24.13 MB


class Solution:
    def numEquivDominoPairs(self, dominoes: List[List[int]]) -> int:
        pair_frequency = {}
        number_of_equivalent_pairs = 0
        for domino in dominoes:
            domino_sorted_order = ()
            if domino[0] < domino[1]:
                domino_sorted_order = (domino[0], domino[1])
            else:
                domino_sorted_order = (domino[1], domino[0])

            frequency = pair_frequency.get(domino_sorted_order)
            if frequency:
                number_of_equivalent_pairs += frequency
                pair_frequency.update({domino_sorted_order: frequency + 1})
            else:
                pair_frequency.update({domino_sorted_order: 1})

        return number_of_equivalent_pairs

File: c++/1128.cpp
// Link: https://leetcode.com/problems/number-of-equivalent-domino-pairs/description/
// Runtime: 4 ms
// Memory: 26.35 MB

class Solution {
public:
  int numEquivDominoPairs(const std::vector<std::vector<int>> &dominoes) {
    std::map<std::pair<int, int>, int> pairFrequency;
    int numberOfEquivalentPairs = 0;
    for (const auto &domino : dominoes) {
      std::pair<int, int> dominoSortedOrder = {};
      if (domino[0] > domino[1])
        dominoSortedOrder = std::make_pair(domino[1], domino[0]);
      else
        dominoSortedOrder = std::make_pair(domino[0], domino[1]);

      int frequency = pairFrequency[dominoSortedOrder];
      if (frequency > 0)
        numberOfEquivalentPairs += frequency;

      pairFrequency[dominoSortedOrder]++;
    }

    return numberOfEquivalentPairs;
  }
};
